adapters/repos/db/vector/flat/metadata.go

```diff
@@ -0,0 +1,161 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package flat
+
+import (
+	"encoding/binary"
+	"fmt"
+	"path/filepath"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	bolt "go.etcd.io/bbolt"
+)
+
+const (
+	metadataPrefix       = "meta"
+	vectorMetadataBucket = "vector"
+)
+
+func (index *flat) getMetadataFile() string {
+	if index.targetVector != "" {
+		// This may be redundant as target vector is already validated in the schema
+		cleanTarget := filepath.Clean(index.targetVector)
+		cleanTarget = filepath.Base(cleanTarget)
+		return fmt.Sprintf("%s_%s.db", metadataPrefix, cleanTarget)
+	}
+	return fmt.Sprintf("%s.db", metadataPrefix)
+}
+
+func (index *flat) initMetadata() error {
+	path := filepath.Join(index.rootPath, index.getMetadataFile())
+	var err error
+	index.metadata, err = bolt.Open(path, 0o600, nil)
+	if err != nil {
+		return errors.Wrapf(err, "open %q", path)
+	}
+	err = index.metadata.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(vectorMetadataBucket))
+		if err != nil {
+			return errors.Wrap(err, "create bucket")
+		}
+		if b == nil {
+			return errors.New("failed to create or get bucket")
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "init metadata bucket")
+	}
+	index.initDimensions()
+	defer func() {
```

Comment:
isn't this too late? if err is !=nil we would not reach this code and the defer function never be executed

Reply:
Thanks, I've moved the defer up